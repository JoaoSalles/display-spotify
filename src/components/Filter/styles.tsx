import styled from 'styled-components';
import { BACKGROUND_WHITE, BORDER_WHITE } from 'constants/colors';

export const FilterElement = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;
`;


export const FilterSearchPainel = styled.div`
    border: 1px solid ${BORDER_WHITE};
    box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.05);
    display: flex;
    margin: 20px 0;
    justify-content: center;
    width: 100%;

    & > div {
        flex: 0 100%;
    }
`;

export const FilterPainel = styled.div`
    background: ${BACKGROUND_WHITE};
    box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.05);
    border: 1px solid ${BORDER_WHITE};
    display: flex;
    flex-wrap: wrap;
    padding: 10px 0;
    width: 100%;

    & > div {
        margin: 20px 20px;
    }

    @media only screen and (max-width: 800px) {
        & > div {
            margin: 10px 0;
        }
    }

    @media only screen and (min-width: 800px) {
        & > div {
            flex: 0 45%;
        }
    }
    

    @media only screen and (min-width: 1200px) {
        & > div {
            flex: 0 45%;
        }
    }
`;