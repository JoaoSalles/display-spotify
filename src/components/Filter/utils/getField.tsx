import React, { ReactElement } from "react";
import { filterFormatField } from 'types/filterFormatTypes';
import FilterInput from 'components/FilterInput';
import FilterInputDataPicker from 'components/FilterInputDataPicker';
import FilterInputSelect from 'components/FilterInputSelect';

export const getField = (
  field: filterFormatField,
  handleFieldChange: ({ field, value }: { field: string, value: string}) => void,
  filterData: {[key: string]: string}
  ): ReactElement => {
  const { id, name } = field;
  let max, min, pattern;
  let values: any = [];

  if (field.validation) {
    max = field.validation.max;
    min = field.validation.min;
    pattern = field.validation.pattern;
  }

  if (field.values) {
    values = field.values
  }

  switch (id) {
    case "locale":
      return (<FilterInputSelect options={values} field={id} name={name} handleFieldChange={handleFieldChange} key={id}/>);

    case "country":
      return (<FilterInputSelect options={values} field={id} name={name} handleFieldChange={handleFieldChange} key={id}/>);

    case "timestamp":
      return (<FilterInputDataPicker filterData={filterData[id]} field={id} name={name} pattern={pattern} handleFieldChange={handleFieldChange} key={id}/>)

    case "limit":
      return (<FilterInput field={id} name={name} min={min} max={max} handleFieldChange={handleFieldChange} key={id}/>);

    case "offset":
      return (<FilterInput field={id} name={name} min={0} handleFieldChange={handleFieldChange} key={id}/>);

    default:
      return (<FilterInput field={id} name={name} handleFieldChange={handleFieldChange} key={id}/>);
  }
};
