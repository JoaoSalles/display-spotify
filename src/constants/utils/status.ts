export const LOADED = "LOADED";
export const PEDDING = "PEDDING";
export const LOADING = "LOADING";
