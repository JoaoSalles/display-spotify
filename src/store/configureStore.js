import { createStore, applyMiddleware, compose } from 'redux'
import { persistStore, persistReducer } from 'redux-persist'
import createSagaMiddleware from 'redux-saga'
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web

import rootReducer from './reducers'
import mySaga from './sagas'

const persistConfig = {
  key: 'root',
  storage,
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

const sagaMiddleware = createSagaMiddleware()

const composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  let store = createStore(
    persistedReducer, 
    composeEnhancer(applyMiddleware(sagaMiddleware)),
  );
  let persistor = persistStore(store);
  sagaMiddleware.run(mySaga);
  return { store, persistor }
}