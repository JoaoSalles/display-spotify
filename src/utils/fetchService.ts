export default async function fetchService(input: RequestInfo, init?: RequestInit):  Promise<Response> {
    return fetch(input, init)
}