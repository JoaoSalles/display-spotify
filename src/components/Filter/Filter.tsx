import React from "react";
import FilterSearch from "components/FilterSearch";
import { Spinner } from "baseui/spinner";
import { getField } from "./utils/getField";
import { FilterElement, FilterPainel, FilterSearchPainel } from "./styles";
import { filterFormatField } from "types/filterFormatTypes";

interface Props {
  filterFields: Array<filterFormatField>,
  filterData: { [key: string]: string},
  handleFieldChange: ({ field, value, }: {
        field: string;
        value: string;
    }) => void,
  loading: boolean
}

function Filter(props: Props) {
  const { filterFields, filterData, handleFieldChange, loading } = props;

  return (
    <FilterElement className="container">
      <FilterPainel className="painel">
        {loading ? (
          <Spinner data-testid="loading" color="#EA1D2C" />
        ) : (
          filterFields.map((filterField) =>
            getField(filterField, handleFieldChange, filterData)
          )
        )}
      </FilterPainel>
      <FilterSearchPainel className="search-painel">
        <FilterSearch name={"Filtrar por nome..."} />
      </FilterSearchPainel>
    </FilterElement>
  );
}

export default Filter;
