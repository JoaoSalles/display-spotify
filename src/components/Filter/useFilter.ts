import { useState, useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux'
import { filterFormatField } from 'types/filterFormatTypes';
import { PLAYLIST_SAGA } from 'store/actions/playlistActions';
import { RootState } from 'store/reducers';
import getURLParams from 'utils/getURLParams';
import * as status from 'constants/utils/status';

const useFilter = () => {
    const dispatch = useDispatch();
    const auth = useSelector((state: RootState) => state.auth);
    const [loading, setLoading] = useState<boolean>(true);
    const [filterFields, setFilterFields] = useState<Array<filterFormatField>>([]);
    const [filterData, setFilterData] = useState<{[key: string]: string}>({});

    const handleFieldChange = ({ field, value }: { field: string, value: string}) => {
        const newValue: {[key: string]: string} = {};
        newValue[field] = value;
        setFilterData({ ...filterData, ...newValue });
      }

    async function getFilterFormat() {
        const response = await fetch('http://www.mocky.io/v2/5a25fade2e0000213aa90776')
        if (response.status === 200) {
            const reponseJson = await response.json();
            setLoading(false);
            setFilterFields(reponseJson.filters);
        }
    }

    async function getPlaylists() {
        if( auth.status === status.LOADED) {
            dispatch({ 
                type: PLAYLIST_SAGA,
                payload: {
                    token: auth.access_token,
                    url: `https://api.spotify.com/v1/browse/featured-playlists${getURLParams(filterData)}`
                }
            });
        }
    }

    useEffect(() => {
        const interval = setInterval(() => {
            getPlaylistsMemo();
        }, 30000);
        return () => clearInterval(interval);
    // eslint-disable-next-line
    }, [auth, filterData]);

    const getPlaylistsMemo = useCallback(
        () => {
            getPlaylists()
        },
        // eslint-disable-next-line
        [auth, filterData]
    );

    useEffect(() => {
        getPlaylistsMemo();
    }, [filterData, getPlaylistsMemo]);

    useEffect(() => {
        getFilterFormat();
    // eslint-disable-next-line
    }, []);

    return {
        filterFields,
        loading,
        handleFieldChange,
        filterData
    }
}

export default useFilter;