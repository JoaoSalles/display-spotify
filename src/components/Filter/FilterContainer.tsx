import React from "react";
import Filter from "./Filter";
import useFilter from "./useFilter";

interface Props {}

function FilterContainer(props: Props) {
  const { filterFields, filterData, handleFieldChange, loading } = useFilter();

  return ( 
    <Filter 
      filterFields={filterFields}
      filterData={filterData}
      handleFieldChange={handleFieldChange}
      loading={loading}
    />
  );
}

export default FilterContainer;
