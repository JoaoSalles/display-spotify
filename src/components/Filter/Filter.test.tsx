import React from "react";
import { render, fireEvent, waitFor, RenderResult } from "@testing-library/react";
import Filter from "./Filter";
import * as status from "constants/utils/status";

import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

describe("Tests for Filter Painel", () => {
    const mockStore = configureStore();
    const defaultInitialState = { auth: { status: status.LOADED }}
    const defaultFilters = [
        {
            id: "limit",
            name: "Quantidade",
            validation: {
                max: 50,
                min: 1,
                primitiveType: "INTEGER"
            }
        },
        {
            id: "offset",
            name: "Página",
            validation: {
                primitiveType: "INTEGER"
            }
        }
    ]

    const defaultProps = {
        filterFields: defaultFilters,
        filterData: {},
        handleFieldChange: () => {},
        loading: true
    }

    const renderFilter = (
        props: { [key: string]: any } = {},
        initialState: { [key: string]: any } = defaultInitialState
    ): RenderResult => {
        let store = mockStore({ ...initialState })
        return render(
            <Provider store={store}><Filter {...defaultProps} {...props} /></Provider>
        );
    }

    it("Should match snapshot", async () => {
        const { container } = renderFilter();

        expect(container).toMatchSnapshot();
    });

    it("Should render loading", async () => {
        const { getByTestId } = renderFilter();

        const buttonField = await waitFor(() => getByTestId("loading"));
        expect(buttonField).toBeTruthy();
    });

    it("Should render Filter inputs", async () => {
        const { getByPlaceholderText } = renderFilter({ loading: false });

        const paginaInput = await waitFor(() => getByPlaceholderText("Página"));

        expect(paginaInput).toBeTruthy();

        const quantidadeInput = await waitFor(() => getByPlaceholderText("Quantidade"));

        expect(quantidadeInput).toBeTruthy();
    });

    it("Should call handleFieldChange on input change", async () => {
        const mockOnchange = jest.fn();
        const { getByPlaceholderText } = renderFilter({ loading: false, handleFieldChange: mockOnchange });

        const paginaInput = await waitFor(() => getByPlaceholderText("Página"));

        fireEvent.change(paginaInput, { target: { value: '23' } });

        await waitFor(() => expect(mockOnchange).toHaveBeenCalledTimes(1));

        expect(mockOnchange).toBeCalled();
    });
});
