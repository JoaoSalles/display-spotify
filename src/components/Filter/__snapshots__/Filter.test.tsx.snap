// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Filter Painel Should match snapshot 1`] = `
<div>
  <div
    class="sc-bdnylx imbnHh container"
  >
    <div
      class="sc-dlnjPT kpeorE painel"
    >
      <svg
        aria-label="Loading"
        class=""
        data-baseweb="spinner"
        data-testid="loading"
        viewBox="3 3 18 18"
      >
        <title>
          Loading
        </title>
        <path
          class=""
          clip-rule="evenodd"
          d="M12 5C8.13401 5 5 8.13401 5 12C5 15.866 8.13401 19 12 19C15.866 19 19 15.866 19 12C19 8.13401 15.866 5 12 5ZM3 12C3 7.02944 7.02944 3 12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12Z"
          fill-rule="evenodd"
        />
        <path
          class=""
          clip-rule="evenodd"
          d="M16.9497 7.05015C14.2161 4.31648 9.78392 4.31648 7.05025 7.05015C6.65973 7.44067 6.02656 7.44067 5.63604 7.05015C5.24551 6.65962 5.24551 6.02646 5.63604 5.63593C9.15076 2.12121 14.8492 2.12121 18.364 5.63593C18.7545 6.02646 18.7545 6.65962 18.364 7.05015C17.9734 7.44067 17.3403 7.44067 16.9497 7.05015Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="sc-gtssRu cmgIYK search-painel"
    >
      <div
        class=""
        data-baseweb="input"
      >
        <div
          class=""
          data-baseweb="base-input"
        >
          <input
            aria-invalid="false"
            aria-required="false"
            autocomplete="on"
            class=""
            inputmode="text"
            name=""
            placeholder="Filtrar por nome..."
            type="text"
            value=""
          />
        </div>
        <div
          class=""
        >
          <svg
            class=""
            data-baseweb="icon"
            title="Search"
            viewBox="0 0 24 24"
          >
            <title>
              Search
            </title>
            <path
              clip-rule="evenodd"
              d="M11 6C8.79086 6 7 7.79086 7 10C7 12.2091 8.79086 14 11 14C13.2091 14 15 12.2091 15 10C15 7.79086 13.2091 6 11 6ZM5 10C5 6.68629 7.68629 4 11 4C14.3137 4 17 6.68629 17 10C17 11.2958 16.5892 12.4957 15.8907 13.4765L19.7071 17.2929C20.0976 17.6834 20.0976 18.3166 19.7071 18.7071C19.3166 19.0976 18.6834 19.0976 18.2929 18.7071L14.4765 14.8907C13.4957 15.5892 12.2958 16 11 16C7.68629 16 5 13.3137 5 10Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
